op-service/txmgr/queue.go

```diff
@@ -0,0 +1,154 @@
+		maxPending = math.MaxInt
+	}
+	return &Queue[T]{
+		txMgr:          txMgr,
+		maxPending:     maxPending,
+		pendingChanged: pendingChanged,
+	}
+}
+
+// Wait waits for all pending txs to complete (or fail).
+func (q *Queue[T]) Wait() {
+	if q.group == nil {
+		return
+	}
+	_ = q.group.Wait()
+}
+
+// Send will wait until the number of pending txs is below the max pending,
+// and then send the next tx. The TxFactory should return an error if the
+// next tx does not exist, which will be returned from this method.
+//
+// The actual tx sending is non-blocking, with the receipt returned on the
+// provided receipt channel.
+func (q *Queue[T]) Send(ctx context.Context, factory TxFactory[T], receiptCh chan TxReceipt[T]) error {
+	ctx, cancel := q.mergeWithGroupContext(ctx)
+	factoryErrCh := make(chan error)
+	q.group.Go(func() error {
+		defer cancel()
+		return q.sendTx(ctx, factory, factoryErrCh, receiptCh)
+	})
+	return <-factoryErrCh
+}
+
+// TrySend sends the next tx, but only if the number of pending txs is below the
+// max pending, otherwise the TxFactory is not called (and no error is returned).
+// The TxFactory should return an error if the next tx does not exist, which is
+// returned from this method.
+//
+// Returns false if there is no room in the queue to send. Otherwise, the
+// transaction is queued and this method returns true.
+//
+// The actual tx sending is non-blocking, with the receipt returned on the
+// provided receipt channel.
+func (q *Queue[T]) TrySend(ctx context.Context, factory TxFactory[T], receiptCh chan TxReceipt[T]) (bool, error) {
+	ctx, cancel := q.mergeWithGroupContext(ctx)
+	factoryErrCh := make(chan error)
+	started := q.group.TryGo(func() error {
+		defer cancel()
+		return q.sendTx(ctx, factory, factoryErrCh, receiptCh)
+	})
+	if !started {
+		cancel()
+		return false, nil
+	}
+	err := <-factoryErrCh
+	return err == nil, err
+}
+
+func (q *Queue[T]) sendTx(ctx context.Context, factory TxFactory[T], factoryErrorCh chan error, receiptCh chan TxReceipt[T]) error {
+	candidate, id, err := factory(ctx)
+	factoryErrorCh <- err
+	if err != nil {
+		// Factory returned an error which was returned in the channel. This means
+		// there is no tx to send, so return nil.
+		return nil
+	}
+
+	q.pendingChanged(q.pending.Add(1))
+	defer func() {
+		q.pendingChanged(q.pending.Add(^uint64(0))) // -1
+	}()
+	receipt, err := q.txMgr.Send(ctx, candidate)
+	receiptCh <- TxReceipt[T]{
+		ID:      id,
+		Receipt: receipt,
+		Err:     err,
+	}
+	return err
+}
+
+// mergeWithGroupContext creates a new Context that is canceled if either the given context is
+// Done, or the group context is canceled. The returned CancelFunc should be called once finished.
+//
+// If the group context doesn't exist or has already been canceled, a new one is created after
+// waiting for existing group threads to complete.
+func (q *Queue[T]) mergeWithGroupContext(ctx context.Context) (context.Context, context.CancelFunc) {
```

Comment:
I have an intuition that we can get rid of `mergeWithGroupContext` if we add a life-time context to the `Queue` struct (would use `l.killCtx` when we initialize it in the batcher) and then derive the errgroup using [`WithContext`](https://pkg.go.dev/golang.org/x/sync/errgroup#WithContext). So we basically flip the order of contexts around and derive the errgroup from that one context always. Since we always pass the `killCtx` to `Send`/`TrySend` anyways, this would have the same end result but without `mergeWithGroupContext`.

Reply:
Good call out, refactored 👍